lib/Service/Totp.php

```diff
@@ -120,6 +120,14 @@ public function validateKey(IUser $user, $key) {
 			$secret = $this->crypto->decrypt($dbSecret->getSecret());
 			/* @phan-suppress-next-line PhanRedefinedClassReference */
 			if ($this->otp->checkTotp(Encoding::base32DecodeUpper($secret), $key, 3) === true) {
+				if ($dbSecret->getVerified() !== true) {
+					// If the 2-factor is enforced, the secret might not be verified
+					// yet because it could come from the challenge page. The user
+					// can't verify the secret and must enter the code directly.
+					// At this point, the challenge has been verified, so we mark
+					// the secret as verified.
+					$dbSecret->setVerified(true);
```

Comment:
would make sense here to check if 2-factor is enforced? or this would get too complicated ?

Reply:
same above

Reply:
Enforcement is done in core, and the providers don't know about it. Right now, the information isn't propagated, and there isn't any specific service in core to get this information from. Note that accessing core's configuration from an app isn't recommended.

For now, it seems better to assume that the challenge page can be shown without the user having configured the app

Reply:
ok, it is safe to assume that starting from 10.13 2-factor can be enforced and logic needs to account for this. what I am worried is what if some app needs to have if/else block and cannot have common handling?

Reply:
Anything more specific? I think the condition is whether the user has configured the 2-factor app or not. If they can know that information, they can move a part of the configuration (whatever they consider relevant) to the challenge page. If they can't it's probably because it's an external service, and the configuration should likely be done in that external service; in this case a link or a reminder could work.